File: implementations/python/charter_verify/base64url.py
"""Unpadded base64url, with one spelling per byte string. SPEC section 8.

"The alphabet is the URL-safe one, `=` is not allowed, and a trailing bit
pattern that is not zero (the classic `AB` versus `AA` ambiguity) is refused:
canonical base64url has one spelling per byte string, and this format uses that
one."

Every way a string can fail to be the one spelling of its bytes carries
NON_CANONICAL_ENCODING, whatever the reason it is not: a character outside the
alphabet, a length no byte string can produce (one leftover character), `=`
padding, and a non-zero trailing bit pattern. SPEC section 5 states the rule —
"the string is not the one spelling this format fixes for that value, whatever
the reason it is not" — and names two of the four cases. MALFORMED is not this
module's to report: a string that is not a byte string of the length its
algorithm requires is the *caller's* complaint, because the length is a fact
about the value and not about the alphabet (section 5's second paragraph).

The empty string is not refused here. It is the one spelling of the empty byte
string, and whether a signature may be one is a question about length, which the
check that reads the field answers with MALFORMED.
"""

from __future__ import annotations

from .errors import Refusal
from .vocabulary import NON_CANONICAL_ENCODING

ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
_INDEX = {character: value for value, character in enumerate(ALPHABET)}


def decode(text: str, what: str = "value") -> bytes:
    """Decode canonical unpadded base64url, or refuse it by name."""
    if "=" in text:
        raise Refusal(
            NON_CANONICAL_ENCODING,
            f"{what} carries '=', and charter/0.1 base64url is unpadded",
        )
    if len(text) % 4 == 1:
        raise Refusal(
            NON_CANONICAL_ENCODING,
            f"{what} is {len(text)} base64url character(s), which no byte string "
            "is spelled with",
        )
    accumulator = 0
    bits = 0
    out = bytearray()
    for character in text:
        value = _INDEX.get(character)
        if value is None:
            raise Refusal(
                NON_CANONICAL_ENCODING,
                f"{what} holds {character!r}, which is not a base64url character",
            )
        accumulator = (accumulator << 6) | value
        bits += 6
        if bits >= 8:
            bits -= 8
            out.append((accumulator >> bits) & 0xFF)
    if bits > 0 and (accumulator & ((1 << bits) - 1)) != 0:
        raise Refusal(
            NON_CANONICAL_ENCODING,
            f"{what} ends on a non-zero trailing bit pattern, so these bytes "
            "have a second spelling",
        )
    return bytes(out)


def encode(data: bytes) -> str:
    """The one base64url spelling of these bytes, without padding."""
    accumulator = 0
    bits = 0
    out = []
    for byte in data:
        accumulator = (accumulator << 8) | byte
        bits += 8
        while bits >= 6:
            bits -= 6
            out.append(ALPHABET[(accumulator >> bits) & 0x3F])
    if bits > 0:
        out.append(ALPHABET[(accumulator << (6 - bits)) & 0x3F])
    return "".join(out)

File: verifier/base64url.js
/**
 * Canonical base64url, decode side.
 *
 * Charter writes keys and signatures as unpadded base64url. Decoding is strict
 * on purpose: unpadded base64url is not a bijection onto byte strings unless
 * the unused trailing bits are required to be zero. Without that rule the same
 * signature has 4 accepted spellings, and "one byte sequence per logical
 * value" would be false for the very field that carries the signature.
 *
 * The empty string is not refused here. It is the one spelling of the empty byte
 * string, and whether a *signature* may be one is a question about length — the
 * caller's, and one it answers MALFORMED rather than NON_CANONICAL_ENCODING.
 *
 * @module verifier/base64url
 */

const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';

const VALUES = (() => {
  const table = new Int8Array(128).fill(-1);
  for (let i = 0; i < ALPHABET.length; i += 1) {
    table[ALPHABET.charCodeAt(i)] = i;
  }
  return table;
})();

/**
 * @param {string} text
 * @returns {{ ok: true, value: Uint8Array } | { ok: false, detail: string }}
 */
export function decodeBase64Url(text) {
  if (text.includes('=')) {
    return { ok: false, detail: 'padding is present; canonical base64url is unpadded' };
  }
  const remainder = text.length % 4;
  if (remainder === 1) {
    return { ok: false, detail: 'the length is 4n+1, which no base64url string can have' };
  }
  const wholeGroups = Math.floor(text.length / 4);
  const byteLength = wholeGroups * 3 + (remainder === 0 ? 0 : remainder - 1);
  const out = new Uint8Array(byteLength);
  let buffer = 0;
  let bits = 0;
  let at = 0;
  for (let i = 0; i < text.length; i += 1) {
    const code = text.charCodeAt(i);
    const value = code < 128 ? VALUES[code] : -1;
    if (value < 0) {
      return { ok: false, detail: `character ${i + 1} is not in the base64url alphabet` };
    }
    buffer = (buffer << 6) | value;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out[at] = (buffer >> bits) & 0xff;
      at += 1;
    }
  }
  if (bits > 0) {
    const leftover = buffer & ((1 << bits) - 1);
    if (leftover !== 0) {
      return { ok: false, detail: 'the unused trailing bits are not zero, so this is one of several spellings of the same bytes' };
    }
  }
  return { ok: true, value: out };
}
